File: utility/encode_step1_grab_metadata_from_link.R
library(jsonlite)
library(RCurl)


### Settings ===================================================

  # The url below pulls up K562 ChIP-seq data of interest from ENCODE

  args <- commandArgs(trailingOnly = TRUE)
  encode.experiment.links <- args[1]
  output.file.samples <- paste(args[2],'samples.tsv',sep="/")
  output.file.libraries <- paste(args[2],'libraries.tsv',sep="/")
  output.file.sequencings <-paste(args[2],'sequencings.tsv',sep="/") 
  
  acceptable.species <- c("human", "mouse", "rat", "cattle")  
  cell.labels <- c("cell line", "primary cell", "in vitro differentiated cells")
  tissue.labels <- c("tissue")
  experiment.types <- c("ATAC-seq",
                        "ChIP-seq")
  
  download.link.prefix <- "https://www.encodeproject.org/files/"
  
### GET METADATA FROM ENCODE DCC ===================================================
  # I use the jsonlite package to query the json format metadata at DCC. 
  # I imagine that users could input a URL like that links directly to their sample(s) of interst. (we should prob do a quick check to make sure that they don;t accidently put a URL that has 1000s of samples! Maybe we cap at 20. If they need more that can use multiple request)

  json <- fromJSON(encode.experiment.links)

  # The first thing I  extract is the experiment  metadata. 
  #We don't have an experiment object in out LIMS, but bascially it is a way to link together different biological replicates (for example, 2 replicates of H3K27ac from K562)
  exp.list.in<-json[["@graph"]]
  exp.list.in$biosample_ontology <- unlist(exp.list.in$biosample_ontology)

  # We still need to extract more metadata about the biosamples, replicates, and sequncing files from these experiments of interest
  # First, biosamples.... ================================

    # initialize table with the sample columns as LIMS template:samples
    template.samples <- as.data.frame(matrix(nrow=0, ncol=25))
    colnames(template.samples)<-c("Date",
                                  "PI",
                                  "Research contact name",
                                  "Research contact e-mail",
                                  "Research contact phone",
                                  "Fiscal contact name",
                                  "Fiscal conact e-mail",
                                  "Index for payment",
                                  "Sample ID",
                                  "Sample description",
                                  "Species",
                                  "Sample type",
                                  "Preperation",
                                  "Fixation?",
                                  "Sample amount",
                                  "Units",
                                  "Service requested",
                                  "Sequencing depth to target",
                                  "Sequencing length requested",
                                  "Sequencing type requested",
                                  "Notes",
                                  "Date sample received",
                                  "Initials of reciever",
                                  "Storage location",
                                  "Internal Notes")
    
    # loop through the experiments to extract biosample info
    # Note that one experiment may have multiple samples because it consists of multiple replicates
    biosample.list <- c()
    for(i in seq(exp.list.in$accession)){
      biosample.list <- c(biosample.list, exp.list.in$replicates[[i]]$library$biosample$accession)
    }
    biosample.list <- unique(biosample.list)
    
    for(b in seq(biosample.list)){
      template.samples[nrow(template.samples)+1,] <- NA
      temp.json<-fromJSON(paste("https://www.encodeproject.org/", biosample.list[b], "/?frame=embedded", sep = ""))
      #add sample ID
      template.samples$`Sample ID`[nrow(template.samples)] <- paste(temp.json$biosample_ontology$term_name, biosample.list[b], sep="_")
      #add sample description
      template.samples$`Sample description`[nrow(template.samples)] <- temp.json$summary
      #add species (only handle human and mouse for now)
      temp.species <- temp.json$organism$name
      if(temp.species %in% acceptable.species){
        template.samples$Species[nrow(template.samples)] <- temp.species
      } else {
        stop("incompatible species!")
      }
      #add sample type (only handle cells and tissue for now)
      temp.type <- temp.json$biosample_ontology$classification
      if(temp.type %in% cell.labels){
        template.samples$`Sample type`[nrow(template.samples)] <- "cultured cells"
      } else {
        if(temp.type %in% tissue.labels){
          template.samples$`Sample type`[nrow(template.samples)] <- "tissue"
        } else {
          stop("incompatible sample type!")
        }
      }
    }
    
    #add constant metadata
    template.samples$Date <- Sys.Date()
    template.samples$PI <- "Gorkin"
    template.samples$Notes <- "pseudosample, data downloaded from ENCODE"
    template.samples[,c("Preperation",
                        "Fixation?",
                        "Units",
                        "Service requested",
                        "Sequencing depth to target",
                        "Sequencing length requested",
                        "Sequencing type requested",
                        "Storage location")] <- "Other (please explain in notes)"
    
    # output sample table
    write.table(template.samples, output.file.samples, quote=F, row.names=F, sep="\t")
  
  # Next, libraries.... ================================
    
    template.libraries <- as.data.frame(matrix(nrow = 0, ncol=10))
    colnames(template.libraries)<-c("Sample ID (Must Match Column I in Sample Sheet)",
                                    "Library description",
                                    "Team member intials",
                                    "Date experiment started",
                                    "Date experiment completed",
                                    "Experiment type",
                                    "Protocol used",
                                    "Reference to notebook and page number",
                                    "Library ID (if library generated)",
                                    "Notes")
    
    library.list <- c()
    for(i in seq(exp.list.in$accession)){
      library.list <- c(library.list, exp.list.in$replicates[[i]]['@id'][,1])
    }
    library.list <- unique(library.list)
    for(l in seq(library.list)){
      template.libraries[nrow(template.libraries)+1,] <- NA
      temp.json<-fromJSON(paste("https://www.encodeproject.org/", library.list[l], "/?frame=embedded", sep = ""))
      #add sample ID
      template.libraries$`Sample ID (Must Match Column I in Sample Sheet)`[nrow(template.libraries)] <- 
        template.samples$`Sample ID`[grep(temp.json$library$biosample$accession, template.samples$`Sample ID`)]
      #add library description
      template.libraries$`Library description`[nrow(template.libraries)] <- paste("ENCODE", temp.json$experiment$description)
      #add experiment type
      temp.assay <- temp.json$experiment$assay_term_name
      if (temp.assay %in% experiment.types){
        template.libraries$`Experiment type`[nrow(template.libraries)] <- temp.json$experiment$assay_term_name
      } else{
        stop("incompatible experiment type")
      }
      #add library ID
      template.libraries$`Library ID (if library generated)`[nrow(template.libraries)] <- temp.json$libraries$accession
    }
    
    #add constant metadata
    template.libraries$Notes <- "pseudolibrary, ENCODE library downloaded from encodeproject.org"
    
    # output sample table
    write.table(template.libraries, output.file.libraries, quote=F, row.names=F, sep="\t")
      
  # Finally, sequencings... ================================
    
    template.seqs <- as.data.frame(matrix(nrow = 0, ncol=17))
    colnames(template.seqs) <- c("Sample ID (Must Match Column I in Sample Sheet)",
                                    "Label (for QC report)",
                                    "species",
                                    "Team member intials",
                                    "Date submitted for sequencing",
                                    "Library ID",
                                    "Sequencing ID",
                                    "Experiment type",
                                    "Sequening core",
                                    "Machine",
                                    "Sequening length",
                                    "Read type",
                                    "Portion of lane",
                                    "i7 index (if applicable)",
                                    "i5 Index (or single index)",
                                    "Notes",
                                    "file_download_URL(s)")
    
    temp.libs <- template.libraries$`Library ID (if library generated)`
    already.added <- c()
    for(l in seq(temp.libs)){
      temp.reps <- fromJSON(paste("https://www.encodeproject.org/", temp.libs[l], "/?frame=embedded", sep = ""))$replicates
      temp.sample.id <-  template.libraries$`Sample ID (Must Match Column I in Sample Sheet)`[l]
      for(r in seq(temp.reps)){
        temp.rep.info <- fromJSON(paste("https://www.encodeproject.org/", temp.reps[r], "/?frame=embedded", sep = ""))
        temp.target <- strsplit(strsplit(temp.rep.info$experiment$target, split="/")[[1]][3], split="-")[[1]][1]
        temp.files <- temp.rep.info$experiment$files
        for(f in seq(temp.files)){
          temp.json <- fromJSON(paste("https://www.encodeproject.org/",temp.files[f], "/?frame=embedded", sep = ""))
          if(temp.json$file_format=="fastq"){
            if(temp.json$library$accession==temp.libs[l] & !(temp.json$accession %in% already.added)){
              template.seqs[nrow(template.seqs)+1,] <- NA
              #add sample ID
              template.seqs$`Sample ID (Must Match Column I in Sample Sheet)`[nrow(template.seqs)] <- temp.sample.id
              #add label for report
              template.seqs$`Label (for QC report)`[nrow(template.seqs)] <-  paste(temp.sample.id, temp.target, temp.json$accession, sep="_")
              #add species 
              template.seqs$species[nrow(template.seqs)] <-  template.samples$Species[grep(temp.sample.id, template.samples$`Sample ID`)][1]
              #add lib ID
              template.seqs$`Library ID`[nrow(template.seqs)] <- temp.json$replicate$library$accession
              #add seq ID 
              template.seqs$`Sequencing ID`[nrow(template.seqs)]  <- paste(template.seqs$`Library ID`[nrow(template.seqs)], 
                                                                           temp.rep.info$technical_replicate_number, sep="_")
              #add experiment
              template.seqs$`Experiment type`[nrow(template.seqs)] <-  
                template.libraries$`Experiment type`[grep(temp.sample.id, template.libraries$`Sample ID (Must Match Column I in Sample Sheet)`)][1]
              #add seq length 
              template.seqs$`Sequening length`[nrow(template.seqs)] <-  temp.json$read_length
              #add read.type
              temp.read.type <- temp.json$run_type
              if(temp.read.type == "paired-ended"){
                template.seqs$`Read type`[nrow(template.seqs)] <- "PE"
              } else {
                template.seqs$`Read type`[nrow(template.seqs)] <- "SE"
              }
              #add notes
              temp.accessions <- temp.json$accession
              if(template.seqs$`Read type`[nrow(template.seqs)] == "PE"){
                temp.accessions <- c(temp.accessions, strsplit(temp.json$paired_with, split="/")[[1]][3])
              }
              template.seqs$Notes[nrow(template.seqs)] <- paste("ENCODE file accession(s)", paste(temp.accessions, collapse=","))
              already.added <- c(already.added, temp.accessions)
              #add download URL
              template.seqs$`file_download_URL(s)`[nrow(template.seqs)] <-
                paste(paste(download.link.prefix,temp.accessions,"/@@download/",temp.accessions,".fastq.gz",sep=""), collapse=",")
            }
          }
        }
      }
    }
    
    # output sample table
    write.table(template.seqs, output.file.sequencings, quote=F, row.names=F, sep="\t")

File: utility/encode_step1_grab_metadata_from_link.py
#!/usr/bin/env python
import argparse
import subprocess
import requests
import json
import datetime
import time
import re
import django
from django.contrib.auth.models import Group,User

os.chdir("../")
basedir = os.path.dirname(os.path.abspath(__file__))
sys.path.append(basedir)
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "epigen_ucsd_django.settings")
django.setup()
from masterseq_app.models import SampleInfo,LibraryInfo,SeqInfo


def main():
    # python encode_step1_grab_metadata_from_link.py -u liyuxin -l "https://www.encodeproject.org/search/?type=Experiment&status=released&assay_title=Histone+ChIP-seq&biosample_ontology.term_name=K562&target.label=H3K4me3&target.label=H3K36me3&target.label=H3K27me3&target.label=H3K27ac&target.label=H3K4me1"
    parser = argparse.ArgumentParser()
    parser.add_argument('-l', help='link from encode')
    parser.add_argument('-u',help = 'the user(username) who requests this ENCODE setqc')

    cell_labels = ['cell line', 'primary cell', 'in vitro differentiated cells']
    tissue_labels = ['tissue']

    url = parser.parse_args().l
    headers = {'accept': 'application/json'}
    response = requests.get(url, headers=headers)
    biosample = response.json()
    graph = biosample['@graph']

    # First, biosamples.... ================================

    start = time.time()
    biosample_list = []
    tosave_list = []
    for x in graph:
        for y in x['replicates']:
            biosample_list.append(str(y['library']['biosample']['accession']))
    print(set(biosample_list))
    for sample in set(biosample_list):
        this_url = "https://www.encodeproject.org/"+sample+"/?frame=embedded"
        response = requests.get(this_url, headers=headers)
        this_sample = response.json()
        if str(this_sample['biosample_ontology']['classification']) in cell_labels:
            this_sample_type = 'cultured cells'
        elif str(this_sample['biosample_ontology']['classification']) in tissue_labels:
            this_sample_type = 'tissue'
        else:
            this_sample_type = 'other (please explain in notes)'
        today = datetime.date.today()
        sample_id=str(this_sample['biosample_ontology']['term_name'])+'_'+sample

        sample_type=this_sample_type
        date=today

        if not SampleInfo.objects.filter(sample_id=sample_id).exists():
            tosave_item = SampleInfo(
                sample_id=sample_id,
                description=str(this_sample['summary']),
                species=str(this_sample['organism']['name']),
                sample_type=this_sample_type,
                date=today,
                group=Group.objects.get(name='David Gorkin'),
                notes='pseudosample, data downloaded from ENCODE',
                )   
            tosave_list.append(tosave_item)
    SampleInfo.objects.bulk_create(tosave_list)
    end = time.time()
    print(end - start)

    # Next, libraries.... ================================
    start = time.time()
    library_list = []
    library_id_list = []
    lib_sam = {}
    tosave_list = []
    ib_new_name = []
    for x in graph:
        for y in x['replicates']:
            library_list.append(str(y['@id']))
    print(set(library_list))
    for library in set(library_list):
        this_url = "https://www.encodeproject.org/"+library+"/?frame=embedded"
        response = requests.get(this_url, headers=headers)
        this_library = response.json()

        this_id=str(this_library['libraries'][0]['accession'])
        notes=';'.join(['pseudolibrary, ENCODE library downloaded from encodeproject.org','ENCODE accession:'+this_id])

        library_id_list.append(this_id)
        lib_sam[this_id] = sampleinfo
        existing_flag = 0
        thissample = SampleInfo.objects,get(sample_id=str(this_library['library']['biosample']['accession']))
        for item in thissample.libraryinfo_set.all():
            if this_id == item.notes.split(';')[1].split(':')[1]:
                lib_new_name[this_id] = item.library_id
                existing_flag = 1
                break;
            else:
                existing_flag = 0
        lib_ids = list(LibraryInfo.objects.values_list('library_id', flat=True))
        maxid = max([int(x.split('_')[1]) for x in lib_ids if x.startswith('ENCODE')])
        libid_new = '_'.join(['ENCODE', str(maxid+1)])
        lib_new_name[libid] = libid_new
        if existing_flag == 0:
            tosave_item = LibraryInfo(
                sampleinfo=thissample,
                library_description=' '.join(['ENCODE',str(this_library['experiment']['description'])]),
                experiment_type=str(this_library['experiment']['assay_term_name']),
                library_id=libid_new,
                notes=notes,
                )
            tosave_list.append(tosave_item)
    LibraryInfo.objects.bulk_create(tosave_list)


    end = time.time()
    print(end - start)
    print(library_id_list)

    # Finally, sequencings... ================================
    start = time.time()
    seq_list = []
    tosave_list = []
    for library in set(library_id_list):
        this_url = "https://www.encodeproject.org/"+library+"/?frame=embedded"
        response = requests.get(this_url, headers=headers)
        this_library = response.json()
        for rep in this_library['replicates']:
            this_rep_url = "https://www.encodeproject.org/"+rep+"/?frame=embedded"
            response = requests.get(this_rep_url, headers=headers)
            this_rep_library = response.json()
            target = re.split('/|-',str(this_rep_library['experiment']['target']))[2]

            default_label='_'.join([lib_sam[library],target])
            libraryinfo=LibraryInfo.objects.get(library_id=lib_new_name[library])

            this_uuid = str(this_rep_library['uuid'])
            notes=';'.join(['ENCODE uuid:'+this_uuid,'ENCODE url for files info:'+this_rep_url])

            existing_flag = 0
            for item in libraryinfo.seqinfo_set.all():
                if this_uuid == item.notes.split(';')[0].split(':')[1]:
                    existing_flag = 1
                    break;
                else:
                    existing_flag = 0
            if existing_flag == 0:
                counts = thislibrary.seqinfo_set.all().count()
                if counts == 0:
                    thisseqid = lib_new_name[library]
                else:
                    thisseqid = lib_new_name[library]+'_'+str(counts+1)
                tosave_item = SeqInfo(
                    default_label=default_label,
                    libraryinfo=libraryinfo,
                    seq_id=thisseqid,
                    notes=notes,
                    )
                tosave_list.append(tosave_item)
        SeqInfo.objects.bulk_create(tosave_list)


    end = time.time()
    print(end - start)


if __name__ == '__main__':
    main()
